README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# redditr

<!-- badges: start -->

[![Travis build
status](https://travis-ci.org/geoffwlamb/redditr.svg?branch=master)](https://travis-ci.org/geoffwlamb/redditr)
[![Coverage
status](https://codecov.io/gh/geoffwlamb/redditr/branch/master/graph/badge.svg)](https://codecov.io/github/geoffwlamb/redditr?branch=master)
<!-- badges: end -->

## Overview

redditr is an R package intended to help obtain content from Reddit by
interfacing with the <a href = "https://github.com/pushshift/api">
Pushshift.io Reddit API</a>.

The immediate scope of this package is to provide functionality for
importing Reddit comment and post data into R. Some additional
functionality is provided for handling time-based information. Aside
from that, redditr will not try to do any general text or data cleaning.
Other, far more established packages are better options for handling
text data once it’s been imported into R.

Let’s see how this goes…

## Installation

Install via devtools:

``` r
devtools::install_github("geoffwlamb/redditr")
```

## Examples

The redditr package’s flagship function,
<code>get\_reddit\_content</code>, takes [Pushshift.io API Search
Parameters](https://github.com/pushshift/api#search-parameters-for-comments)
as arguments and returns a data.frame with information related your
query. Below are some ideas for how you can use this function.

### Basic Usage

If you call <code>get\_reddit\_content</code> without specifying any of
the parameters the Pushshift API is looking for, you’ll end up getting
the 500 most recent comments available from the api. If you want
information relating to posts instead, you can change the
<code>content\_type</code> argument to “submission”.

``` r
# load redditr
library(redditr)

# get 500 most recent reddit comments avilable from api
recent_comments <- get_reddit_content()

# get 500 most recent posts
recent_posts <- get_reddit_content(content_type = "submission")
```

### Adjusting Limits to Returned Results and Query Time

The Pushshift API limits returns a maximum of 500 results in a single
query. You can use <code>get\_reddit\_content</code> to automate the
process of sending multiple queries and collecting them into a single
data frame. Also, if you are having issues with queries taking too long
to return information, you can adjust the amount of time the function
will wait for data before giving up. The function defaults to 10
seconds.

``` r
# get more than 500 comments
many_recent_comments <- get_reddit_content(
  content_type = "comment", 
  result_limit = 1000
)

# wait 20 seconds per query
patient_query <- get_reddit_content(
  content_type = "comment",
  timeout = 20
)
```

### Using Pushshift API Parameters

This section is a very basic introduction to including Pushshift API
parameters in your function calls. To explore more complex querying with
the Pushshift API, I highly recommend checking out the [Pushshift
Documentation](https://media.readthedocs.org/pdf/reddit-api/beta/reddit-api.pdf).
In theory, <code>get\_reddit\_content</code> should be able to support
any of the parameters mentioned in the linked documentation. However,
that claim has not been fully verified in practice and only a handful of
unit tests have been written to determine how parameters affect query
results. If you do find any discrepancies between examples in the linked
Pushshift Documentation and function output, please feel free to open an
issue.

#### Using the Search (q) Parameter

The <code>q</code> parameter lets you search for specific text within a
comment or a submission. Here are some use cases for how you might use
it with a multi-word phrase in different ways:

``` r
# get comments containing the string "data science"
# note the double quotes inside the single quotes
data_science_comments <- get_reddit_content(
  content_type = "comment",
  q = '"data science"'
)

# get comments containing the string "data" AND the (separate) string "science"
data_and_science_comments <- get_reddit_content(
  content_type = "comment",
  q = "data+science"
)

# get comments containing the string "data" OR the (separate) string "science"
data_or_science_comments <- get_reddit_content(
  content_type = "comment",
  q = "data|science"
)

# get comments containing the string "data" but NOT the string "science"
# based on some light testing, the parentheses are needed on the non-negated part
# "(data)-science" and "(data)-(science)" do the same thing
# "data-(science)" does NOT
data_not_science_comments <- get_reddit_content(
  content_type = "comment",
  q = "(data)-science"
)

# a more complex query: "data science" or "macine learning" without "statistics"
ds_or_ml_without_stats <- get_reddit_content(
  content_type = "comment",
  q = '("data science"|"machine learning")-statistics'
)
```

#### Specifying Dates in API Queries

There are a few parameters in the Pushshift API that can be used to
filter results based on time. The most common ones are
<code>before</code> and <code>after</code>. These parameters are
expecting dates to be provided in a very particular format: [Unix
Time](https://en.wikipedia.org/wiki/Unix_time). A function for
converting typical date formats to Unix time is available as part of
redditr: <code>date\_to\_api</code>.

``` r
# get comments before a specific date
comments_before_christmas <- get_reddit_content(
  content_type = "comment",
  before = date_to_api("2018-12-25 00:00:00", tz = "EST")
)

# get comments after a specific date 
comments_after_christmas <- get_reddit_content(
  content_type = "comment",
  after = date_to_api("2018-12-25 23:59:59", tz = "EST")
)
```

#### Using Other Pushshift API Parameters

Similar to above, the general format for declaring Pushshift API
parameters in <code>get\_reddit\_content</code> is param = “value”.
Please refer to Pushshift documentation for the full list of known
parameters. Here are a few more examples with some available parameters
that may be of interest:

``` r
# get posts from a specific subreddit
rstats_posts <- get_reddit_content(
  content_type = "submission",
  subreddit = "rstats"
)

# get comments from a specific user
hadley_comments <- get_reddit_content(
  content_type = "comment",
  author = "hadley"
)

# get posts that have received a particular amount of karma
good_posts <- get_reddit_content(
  content_type = "submission",
  score = ">1000"
)

#  combine parameters from all of the sections
data_science_posts_on_rstats_before_christmas <- get_reddit_content(
  content_type = "submission",
  result_limit = 100,
  q = "data science",
  subreddit = "rstats",
  before = date_to_api("2018-12-25 00:00:00", tz = "EST")
)
```

<br>

<hr>

<br>

Hopefully that captures the essence of what this package aims to
accomplish. Please feel free to let me know if anything isn’t working
well and thanks for checking out redditr\!


NEWS.md
# redditr 0.1.1


* Fixed how spacing works when q argument is passed to construct_pushshift_url. It now aligns with pushshift api query documentation.
* Added examples to documentation for using q argument on multi-word terms
* Added more specific unit tests for data importing
* Minor additional documentation updates

# redditr 0.1.0

* Reworked package functions to support a single flagship function, get_reddit_content
* Renamed underlying functions to be more specific
* New unit tests for existing functionality

# redditr 0.0.2

* Added unit tests for interactions between url constructing and data importing
* Added functionality for importing data from constructed urls

# redditr 0.0.1

* Added functionality for constructing URLs to access data via pushshift api

# redditr 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(redditr)

test_check("redditr")


tests/testthat/test_api_params.R
context("each api parameter affects results")

test_that("reference dataset is consistent", {
  a <- get_content(construct_url(size = 10, before = date_to_api("2017-01-01")))
  Sys.sleep(3)
  b <- get_content(construct_url(size = 10, before = date_to_api("2017-01-01")))
  expect_identical(a, b)
})

test_that("known parameters alter comment queries", {
  # baseline, not checking size, before, after, aggs, frequency, or sort_type here
  ref_url <- construct_url(size = 10, before = date_to_api("2016-01-01"))
  ref <- get_content(ref_url)

  expect_false(
    identical(ref, get_content(paste0(ref_url, "&q=programming")))
  )
  expect_false(
    identical(ref, get_content(paste0(ref_url, "&ids=123,456")))
  )
  expect_false(
    identical(ref, get_content(paste0(ref_url, "&fields=created_utc,subreddit")))
  )
  expect_false(
    identical(ref, get_content(paste0(ref_url, "&sort=asc")))
  )
  expect_false(
    identical(ref, get_content(paste0(ref_url, "&author=hadley")))
  )
})

test_that("known parameters alter submission queries", {
  # baseline, not checking size, before, after, aggs, frequency here
  ref_url <- construct_url(
    size = 10,
    before = date_to_api("2016-01-01"),
    type = "submission"
  )
  ref <- get_content(ref_url)

  expect_false(
    identical(ref, get_content(paste0(ref_url, "&q=programming")))
  )
  # These are known issues on the pushshift api gihub
  # expect_false(
  #   identical(ref, get_content(paste0(ref_url, "&ids=123,456")))
  # )
  # expect_false(
  #   identical(ref, get_content(paste0(ref_url, "&fields=created_utc,subreddit")))
  # )
  expect_false(
    identical(ref, get_content(paste0(ref_url, "&sort=asc")))
  )
  expect_false(
    identical(ref, get_content(paste0(ref_url, "&sort_type=score")))
  )
})


tests/testthat/test-url-building.R
context("test-url-building")

test_that("error handling produces expected output", {
  expect_error(
    construct_pushshift_url(content_type = "bad"),
    regexp = 'Invalid content_type argument. Please use either "comment" or "submission"',
    fixed = TRUE
  )
  expect_error(
    construct_pushshift_url(content_type = c("comment", "submission")),
    regexp = 'Please use a single string (either "comment" or "submission") for type',
    fixed = TRUE
  )
  expect_error(
    construct_pushshift_url(content_type = c("bad", "comment")),
    regexp = 'Please use a single string (either "comment" or "submission") for type',
    fixed = TRUE
  )
})

test_that("urls are constructed appropriately", {
  expect_identical(
    construct_pushshift_url(content_type = "comment"),
    "https://api.pushshift.io/reddit/search/comment/?&="
  )
  expect_identical(
    construct_pushshift_url(content_type = "submission"),
    "https://api.pushshift.io/reddit/search/submission/?&="
  )
  expect_identical(
    construct_pushshift_url(content_type = "submission", q = '"hello world"'),
    'https://api.pushshift.io/reddit/search/submission/?&q="hello%20world"'
  )
  expect_identical(
    construct_pushshift_url(content_type = "comment", newparam = "newval"),
    "https://api.pushshift.io/reddit/search/comment/?&newparam=newval"
  )
  expect_identical(
    construct_pushshift_url(content_type = "comment", param1 = "val1", param2 = "val2"),
    "https://api.pushshift.io/reddit/search/comment/?&param1=val1&param2=val2"
  )
  expect_identical(
    construct_pushshift_url(content_type = "comment", param1 = 1, param2 = 2),
    "https://api.pushshift.io/reddit/search/comment/?&param1=1&param2=2"
  )
  expect_identical(
    construct_pushshift_url(content_type = "comment", param1 = 1, param2 = NULL),
    "https://api.pushshift.io/reddit/search/comment/?&param1=1"
  )
})


R/url-building.R
#' Construct a URL to Access Reddit Content
#' @description This function is designed to convert query parameters supplied
#'     as arguments into a URL leading to JSON content that can be accessed
#'     via other functions. In particular, this function is designed to provide
#'     the input for \link{import_reddit_content_from_url}, which, in turn, is
#'     a function that is used to power redditr's flagship function,
#'     \link{get_reddit_content}. Using this function directly may be useful
#'     if you need to access the raw JSON for your query.
#' @param content_type A string containing the type of content you want to
#'     query. The Pushshift API supports the following options: "comment" and
#'     "submission". This function defaults to "comment".
#' @param ... Additional parameters to pass to api query. See Details below.
#' @return A string containing a URL pointing to JSON Reddit data
#' @export
#' @details Here's the list of parameters that the API can use:
#'      https://github.com/pushshift/api#search-parameters-for-comments.
#'      The api ignores unrecognized parameters, so you may want to verify
#'      your output.
#' @examples
#' # url for most recent comments on Reddit
#' construct_pushshift_url(content_type = "comment")
#' 
#' # url for most recent submissions on Reddit
#' construct_pushshift_url(content_type = "submission")
#' 
#' # url for 500 most recent comments on Reddit
#' construct_pushshift_url(size = 500)
#' 
#' # url for 25 most recent posts from the rstats subreddit
#' construct_pushshift_url(type = "submission", subreddit = "rstats")
#' 
#' # url for 25 most recent posts from the rstats subreddit
#' construct_pushshift_url(
#'   type = "submission",
#'   subreddit = "rstats",
#'   before = date_to_api("2017-03-14"),
#'   after = date_to_api("2017-03-12")
#' )
construct_pushshift_url <- function(content_type = "comment", ...) {
  # validate type
  if (length(content_type) != 1) {
    stop('Please use a single string (either "comment" or "submission") for type')
  }
  if (!(content_type %in% c("comment", "submission"))) {
    stop('Invalid content_type argument. Please use either "comment" or "submission"')
  }

  # build stub
  url_stub <- "https://api.pushshift.io/reddit/search/"

  # add type
  type <- paste(content_type, "/?", sep = "")

  # add any other supplied components
  params <- c(...)
  api_query <- paste("&", names(params), "=", params, sep = "", collapse = "")

  # combine parts and replace spaces with pluses
  url_string <- gsub(" ", "%20", paste0(url_stub, type, api_query))

  return(url_string)
}


R/url_building.R
#' Construct a URL to access Reddit Content
#'
#' @param type Type of content you're looking for. Permissible values are "comment" and "submission"
#' @param ... Additional parameters to pass to api query. See details below.
#'
#' @return A string containing a URL pointing to JSON Reddit data
#' @export
#'
#' @details Here's the list of parameters that the api can use: https://github.com/pushshift/api#search-parameters-for-comments. The api ignores unrecognized parameters, so you may want to verify your output.
#'
#' @examples
#' # URL for most recent comments on Reddit
#' construct_url()
#'
#' # Grab 500 instead of default 25
#' construct_url(size = 500)
#'
#' # Grab posts from specific subreddit
#' construct_url(type = "submission", subreddit = "rstats")
#'
#' # Grab posts on a certain date
#' construct_url(
#'  type = "submission,
#'  subreddit = "rstats",
#'  before = date_to_api("2017-03-14"),
#'  after = date_to_api("2017-03-12")
#' )
#'
construct_url <- function(type = "comment", ...) {

  # validate type
  if (length(type) != 1) {
    stop('Please use a single string (either "comment" or "submission") for type')
  }
  if (!(type %in% c("comment", "submission"))) {
    stop('Invalid type argument. Please use either "comment" or "submission"')
  }

  # build stub
  url_stub <- "https://api.pushshift.io/reddit/search/"

  # add type
  type <- paste(type, "/?", sep = "")

  # add any other supplied components
  params <- c(...)
  api_query <- paste("&", names(params), "=", params, sep = "", collapse = "")

  # return URL
  return(paste(url_stub, type, api_query, sep = ""))
}


#' Convert String or Date to API Input
#'
#' @param date A date to compare to the Unit epoch, granular down to seconds
#' @param tz The timezone of the date to compare, defaults to system's timezone
#'
#' @return A string representing a time (in seconds) that the api can use to return results
#' @export
#' @description Converts a supplied date to microseconds, which can get passed to date arguments (e.g. before, after) in construct_url
#' @examples
#' # Convert today
#' date_to_api(Sys.Date())
#'
#' # Pass as argument to construct_url
#' construct_url(before = date_to_api("2017-01-01"))
#'
date_to_api <- function(date, tz = "") {
  origin <- as.POSIXct("1970-01-01", tz = "UTC")
  comp_date <- as.POSIXct(date, tz = tz)

  raw_text <- format(
    as.numeric(difftime(comp_date, origin, units = "secs")),
    scientific = FALSE
  )

  sub("^\\s+", "", raw_text)
}

#' Convert API Time Back to a Timestamp
#' @description This is a convenience function that wraps as.POSIXct to
#'     convert a time represented in seconds after the unit epoch to
#'     a more human interpretable timestamp.
#' @param api_date A date or vector of dates to be converted
#' @param tz The timezone you want the resulting output to be in, defaults
#'     to system's timezone
#' @return A POSIXct of the input
#' @export
#' @examples
#' # Convert current time to API
#' right_now <- Sys.time()
#'
#' # Convert it back to
#' right_now_as_api <- date_to_api(right_now)
#'
#' # And back to current time
#' api_to_date(right_now_as_api)
#'
api_to_date <- function(api_date, tz = "") {
  as.POSIXct(as.numeric(api_date), origin = "1970-01-01", tz = tz)
}


R/construct_url.R
#' Construct a URL to access Reddit Content
#'
#' @param type Type of content you're looking for. Permissible values are "comment" and "submission"
#' @param author String representing the handle of a specific Reddit account
#' @param subreddit String representing a subreddit to extract data from
#' @param size How many records to return, max is 500
#' @param before Integer represting time in . Defaults to current time. Functionality for using this parameter more comprehensively in development.
#' @param manual String that lets you interface directly with the API. Overrides any other parameters and appends passed argument to URL.
#'
#' @return A string representing a URL providing Reddit content
#' @export
#'
#' @examples
#' #Grab 500 most recent comments on Reddit
#' construct_url()
construct_url <- function(
  type = "comment",
  author = "",
  subreddit = "",
  size = 500,
  before = as.numeric(Sys.time()),
  manual = NULL
){

  #validate type
  if (length(type) != 1) {
    stop(
      'Please use a single string (either "comment" or "submission") for type'
    )
  }
  if (!(type %in% c("comment", "submission"))) {
    stop(
      'Invalid type argument. Please use either "comment" or "submission"'
    )
  }

  #build stub
  url_stub <- "https://api.pushshift.io/reddit/search/"

  #add type
  type <- paste(type, "/?", sep = "")

  #override other parameters if manual argument is supplied
  if (!is.null(manual)) {
    return(paste(url_stub, type, manual, sep = ""))
  }

  #add any other supplied components
  before <- paste("before=", round(before), sep = "")
  size <- paste("&size=", size, sep = "")
  if(author != ""){
    author <- paste("&author=", author, sep = "")
  }
  if(subreddit != ""){
    subreddit <- paste("&subreddit=",subreddit, sep = "")
  }

  #return URL
  return(paste(url_stub, type, before, size, author, subreddit, sep = ""))

}


R/data-importing.R
#' Import Data from a Pushshift URL
#' @description This function is a wrapper around \code{\link[jsonlite]{fromJSON}}
#'     that returns a data.frame instead of a list. This function is primarily
#'     useful for powering \link{get_reddit_content}, but can also be called
#'     directly on a single url obtained via \link{construct_pushshift_url}.
#' @param url A url leading to JSON data, typically produced via
#'     \link{construct_pushshift_url}.
#' @param timeout An integer representing the maximum amount of time to allow
#'     for retrieving content from a single URL. Defaults to 10 seconds.
#' @return A data.frame with data available from the provided url
#' @export
#' @examples
#' # read the 25 recent Reddit comments into R
#' import_reddit_content_from_url(construct_pushshift_url())
#' 
#' # construct a url using pushshift api parameters
#' comments_url <- construct_pushshift_url(
#'   content_type = "comment",
#'   q = "Stardew Valley",
#'   subreddit = "games"
#' )
#' 
#' # import data from that url
#' comments_df <- import_reddit_content_from_url(comments_url)
import_reddit_content_from_url <- function(url, timeout = 10) {
  tryCatch(
    expr = R.utils::withTimeout(
      expr = {
        jsonlite::fromJSON(url)[[1]] %>%
          # only keep standard data.frame column types in output
          purrr::keep(~ (is.numeric(.x) | is.character(.x) | is.logical(.x)))
      },
      timeout = timeout
    ),
    TimeoutException = function(ex) {
      message("Time limit for a single request was exceeded.")
    }
  )
}


R/data_importing.R

#' Import data from URL
#'
#' @description This function is a wrapper around jsonlite::fromJSON that returns
#'     a data.frame instead of a list.
#' @param url A url leading to JSON data, typically produced from construct_url
#' @return A data.frame with scraped data from the provided url
#' @export
#' @examples
#' # Read 25 recent Reddit comments into R
#' get_content(construct_url())
get_content <- function(url) {
  # removing nested data for now so that combine_content works as expected
  jsonlite::fromJSON(url)[[1]] %>%
    purrr::modify_if(is.list, function(x) NULL)
}


#' Scrape content from multiple URLs and combine into a single data frame
#' @description This function dynamically constructs URLs and imports the
#'     data contained within them. It is designed to automate the process of importing
#'     larger amounts of data.
#' @param n_max Max number of rows to return. Please keep in mind your available
#'     system resources and any potential burden on other servers when determing
#'     the number of rows you need.
#' @param ... Additional arguments to pass to construct_url to build api query.
#' @return A data.frame with content imported from your query
#' @export
combine_content <- function(n_max, ...) {
  dots <- list(...)
  names(dots) <- tolower(names(dots))
  init_before <- date_to_api(Sys.time())
  init_after <- date_to_api("1970-01-01")
  if ("size" %in% names(dots)) {
    warning("Size cannot be used this function, using specified n_max instead.")
    dots[["size"]] <- NULL
  }
  if ("before" %in% tolower(names(dots))) {
    init_before <- dots[["before"]]
    dots[["before"]] <- NULL
  }
  if ("after" %in% tolower(names(dots))) {
    init_after <- dots[["after"]]
    dots[["after"]] <- NULL
  }

  # initialize variables
  progress_bar <- txtProgressBar(
    min = 0,
    max = n_max,
    style = 3,
    width = 50
  )
  progress <- 0
  size <- min(n_max, 500)
  tmp_left <- n_max
  tmp_url <- do.call(
    construct_url,
    c(
      list(
        size = min(tmp_left, 500),
        before = init_before
      ),
      dots
    )
  )
  tmp_df <- data.frame()

  while (tmp_left > 0) {
    # add new data
    tmp_data <- get_content(tmp_url)
    # hotfix to coerce edited to character
    if ("edited" %in% colnames(tmp_data)) {
      tmp_data$edited <- as.integer(tmp_data$edited)
    }
    tmp_df <- dplyr::bind_rows(tmp_df, tmp_data)

    # update tmp variables
    tmp_left <- n_max - nrow(tmp_df)
    progress <- progress + min(tmp_left, 500)
    tmp_url <- do.call(
      construct_url,
      c(
        list(
          before = min(tmp_df$created_utc),
          size = min(size, tmp_left)
        ),
        dots
      )
    )
    setTxtProgressBar(progress_bar, progress)

    # break if final url
    if (length(tmp_data) == 0) {
      break
    }
    if (nrow(tmp_data) < size) {
      break
    }
  }

  setTxtProgressBar(progress_bar, n_max)
  out <- tmp_df %>%
    dplyr::filter(as.numeric(created_utc) > as.numeric(init_after)) %>%
    purrr::modify_if(grepl("utc", colnames(.)), api_to_date, tz = "UTC")
  return(out)
}


R/date-conversion.R
#' Convert String or Date to API Input
#' @description This function converts dates from a human-consumable format
#'     (e.g. 2008-12-15) to their equivalent time in seconds since the Unit
#'     Epoch (1970-01-01 00:00:00 UTC), which is how Pushshift's API expects
#'     dates.
#' @param date A date (which can be a string representing a validly formatted
#'     date or one of the Date types) to compare to the Unit epoch, granular
#'     down to seconds.
#' @param tz The timezone of the date to compare, defaults to system's timezone
#'     for direct use.
#' @return A string representing a time (in seconds) that the api can use to
#'     return results.
#' @export
#' @examples
#' # convert today
#' date_to_api(Sys.Date())
#' 
#' # convert a specific date
#' date_to_api("2018-03-21")
#' 
#' # pass as argument to construct_pushshift_url
#' construct_pushshift_url(before = date_to_api("2017-01-01"))
date_to_api <- function(date, tz = "") {
  origin <- as.POSIXct("1970-01-01", tz = "UTC")
  comp_date <- as.POSIXct(date, tz = tz)

  raw_text <- format(
    as.numeric(difftime(comp_date, origin, units = "secs")),
    scientific = FALSE
  )

  return(sub("^\\s+", "", raw_text))
}

#' Convert API Time Back to a Timestamp
#' @description This is a convenience function that wraps \code{as.POSIXct} to
#'     convert a time represented in seconds after the unit epoch to
#'     a more human interpretable timestamp.
#' @param api_date A date or vector of dates to be converted
#' @param tz The timezone you want the resulting output to be in, defaults
#'     to system's timezone when calling this function directly.
#' @return A POSIXct of the input
#' @export
#' @examples
#' # capture the current time
#' right_now <- Sys.time()
#' 
#' # convert it to api format (seconds after unit epoch)
#' right_now_as_api <- date_to_api(right_now)
#' 
#' # and back to current time
#' api_to_date(right_now_as_api)
api_to_date <- function(api_date, tz = "") {
  as.POSIXct(as.numeric(api_date), origin = "1970-01-01", tz = tz)
}


R/get-reddit-content.R
#' Scrape Content from Multiple URLs and Combine into a Single Data Frame
#' @description This is the flagship function of redditr. It is designed to
#'     handle the process of constructing a query, generating URLs that
#'     point to content, and importing that content into R. If lower-level
#'     control over that workflow is needed, please see
#'     \link{construct_pushshift_url} and \link{import_reddit_content_from_url}.
#' @param content_type A string containing the type of content you want to
#'     query. The pushshift api supports the following options: "comment" and
#'     "submission". This function defaults to "comment" and gets
#'     passed to \link{construct_pushshift_url}.
#' @param result_limit An integer representing the maximum number of results to
#'     return. Defaults to 500, which is the maximum number of results that can be
#'     returned in a single pushshift api query URL. Please keep in mind your
#'     available system resources and any potential burden on other servers when
#'     determining the number of rows you need.
#' @param timeout An integer representing the maximum amount of time to allow
#'     for retrieving content from a single URL. Defaults to 10 seconds. When
#'     \code{result_limit} is over 500, the timeout resets for every 500 results
#'     that have been returned successfully.
#' @param ... Additional arguments to pass to \link{construct_pushshift_url} that
#'     are used to build the api query.
#' @return A data.frame with content imported from your query
#' @export
#' @examples
#' # basic examples ----
#' 
#' # get 500 most recent reddit comments avilable from api
#' recent_comments <- get_reddit_content()
#' 
#' # get 500 most recent posts
#' recent_posts <- get_reddit_content(content_type = "submission")
#' 
#' # get more than 500 comments
#' many_recent_comments <- get_reddit_content(
#'   content_type = "comment",
#'   result_limit = 1000
#' )
#' 
#' # wait longer than default 10 seconds per query
#' patient_query <- get_reddit_content(
#'   content_type = "comment",
#'   timeout = 20
#' )
#' 
#' 
#' # search term examples ----
#' 
#' # get comments containing the string "data science"
#' # note the double quotes inside the single quotes
#' data_science_comments <- get_reddit_content(
#'   content_type = "comment",
#'   q = '"data science"'
#' )
#' 
#' # get comments containing the string "data" AND the (separate) string "science"
#' data_and_science_comments <- get_reddit_content(
#'   content_type = "comment",
#'   q = "data+science"
#' )
#' 
#' # get comments containing the string "data" OR the (separate) string "science"
#' data_or_science_comments <- get_reddit_content(
#'   content_type = "comment",
#'   q = "data|science"
#' )
#' 
#' # get comments containing the string "data" but NOT the string "science"
#' # based on some light testing, the parentheses are needed on the non-negated part
#' # "(data)-science" and "(data)-(science)" do the same thing
#' # "data-(science)" does NOT
#' data_not_science_comments <- get_reddit_content(
#'   content_type = "comment",
#'   q = "(data)-science"
#' )
#' 
#' # a more complex query: "data science" or "macine learning" without "statistics"
#' ds_or_ml_without_stats <- get_reddit_content(
#'   content_type = "comment",
#'   q = '("data science"|"machine learning")-statistics'
#' )
#' 
#' 
#' # time-based examples ----
#' 
#' # get comments before a specific date
#' comments_before_christmas <- get_reddit_content(
#'   content_type = "comment",
#'   before = date_to_api("2018-12-25 00:00:00", tz = "EST")
#' )
#' 
#' # get comments after a specific date
#' comments_after_christmas <- get_reddit_content(
#'   content_type = "comment",
#'   after = date_to_api("2018-12-25 23:59:59", tz = "EST")
#' )
#' 
#' 
#' # other pushift api parameter examples ----
#' 
#' # get posts from a specific subreddit
#' rstats_posts <- get_reddit_content(
#'   content_type = "submission",
#'   subreddit = "rstats"
#' )
#' 
#' # get comments from a specific user
#' hadley_comments <- get_reddit_content(
#'   content_type = "comment",
#'   author = "hadley"
#' )
#' 
#' # get posts that have received a particular amount of karma
#' good_posts <- get_reddit_content(
#'   content_type = "submission",
#'   score = ">1000"
#' )
#' 
#' #  combine parameters
#' data_science_posts_on_rstats_before_christmas <- get_reddit_content(
#'   content_type = "submission",
#'   result_limit = 100,
#'   q = "data science",
#'   subreddit = "rstats",
#'   before = date_to_api("2018-12-25 00:00:00", tz = "EST")
#' )
get_reddit_content <- function(
                               content_type = "comment",
                               result_limit = 500,
                               timeout = 10,
                               ...) {
  # handle supplied args
  dots <- list(...)
  names(dots) <- tolower(names(dots))

  # warn if size was specified
  if ("size" %in% names(dots)) {
    warning(
      paste0(
        "The size parameter cannot be used with this function.",
        " Using the specified result_limit instead ",
        "(", result_limit, ")."
      )
    )
    dots[["size"]] <- NULL
  }

  # establish maximum time window
  init_before <- date_to_api(Sys.time())
  init_after <- date_to_api("1970-01-01")

  # adjust time window if before or after were supplied
  if ("before" %in% tolower(names(dots))) {
    init_before <- dots[["before"]]
    dots[["before"]] <- NULL
  }
  if ("after" %in% tolower(names(dots))) {
    init_after <- dots[["after"]]
    dots[["after"]] <- NULL
  }

  # initialize working vars
  size <- min(result_limit, 500)
  tmp_left <- result_limit
  tmp_url <- do.call(
    what = construct_pushshift_url,
    args = c(
      list(
        content_type = content_type,
        size = min(tmp_left, 500),
        before = init_before
      ),
      dots
    )
  )
  tmp_df <- data.frame()
  pb_ticks <- ceiling(result_limit / 500)
  pb <- dplyr::progress_estimated(pb_ticks)

  while (tmp_left > 0) {
    # attempt to access new data
    tryCatch(
      expr = {
        tmp_data <- import_reddit_content_from_url(tmp_url, timeout = timeout)
      },
      error = function(cond) {
        warning(
          paste(
            "An error occurred while fetching content.",
            "Returning the current accumulation of content and stopping the process.",
            "Error message:",
            cond
          )
        )
        return({
          tmp_df %>%
            dplyr::filter(as.numeric(created_utc) > as.numeric(init_after)) %>%
            purrr::modify_if(grepl("utc", colnames(.)), api_to_date, tz = "UTC")
        })
      }
    )

    # coerce edited column to character
    if ("edited" %in% colnames(tmp_data)) {
      tmp_data$edited <- as.character(tmp_data$edited)
    }
    # convert retrieved on column to date
    if ("retrieved_on" %in% colnames(tmp_data)) {
      tmp_data$retrieved_on <- api_to_date(tmp_data$retrieved_on, "UTC")
    }

    # add rows retrieved from most recent query to accumulated results
    tmp_df <- dplyr::bind_rows(tmp_df, tmp_data)

    # update progress to output df
    pb$tick()$print()

    # break if nothing returned or final url
    if (is.null(tmp_data)) {
      break
    }
    if (nrow(tmp_data) < size) {
      message("\nReached the end of query results before result limit was met.")
      break
    }

    # update tmp variables
    tmp_left <- result_limit - nrow(tmp_df)
    tmp_url <- do.call(
      what = construct_pushshift_url,
      args = c(
        list(before = min(tmp_df$created_utc), size = min(size, tmp_left)),
        dots
      )
    )
  }

  # filter out records outside of time window and convert times for utc cols
  out <- tmp_df %>%
    dplyr::filter(as.numeric(created_utc) > as.numeric(init_after)) %>%
    purrr::modify_if(grepl("utc", colnames(.)), api_to_date, tz = "UTC")

  # return results
  return(out)
}
